tags/TagMapContext.md

```diff
@@ -0,0 +1,119 @@
+# Summary
+A `Tag` is a key-value pair that can be used to label anything that is associated
+with a specific operation, such as an HTTP request. These `Tag`s are used to
+aggregate measurements in a [`View`](https://github.com/census-instrumentation/opencensus-specs/blob/master/stats/DataAggregation.md#view) 
+according to unique value of the `Tag`s. The `Tag`s can also be used to filter (include/exclude)
+measurements in a `View`. `Tag`s can further be used for logging and tracing.
+
+# TagMap 
+`TagMap` is an abstract data type that represents collection of tags. 
+i.e., each key is associated with exactly one value, but multiple keys can be
+associated with the same value.  `TagMap` is serializable, and it represents
+all of the information that could be propagated inside the process and across process boundaries.  
+`TagMap` is a recommended name but languages can have more language specific name.
+
+## TagKey
+
+`TagKey` consists of source, metadata and name.
+
+### Source
+
+Source of a tag is either remote or local.
+ 
+**Local Source**
+If a tag is created within an entity it is considered local tag for the entity.
+The entity here could be a process, a collection of process, a micro-service, or any arbitrary 
+definition. When a tag is created it's source is always designated as a local tag.
+
+**Remote Source**
+If a tag is received from another entity it is considered as remote tag for the receiving entity.
+
+
+#### Source Transition 
+A receiver at the edge of an entity converts source of all incoming tags to remote tags. Once the
+source of a tag is designated as remote it stays as remote for the rest of the request processing.
+This is applicable to reentrant processes as well ie. A tag T is created in an entity E and goes
+through process P as local tag. It then leaves entity E and 
+returns back and goes through Process P again. Tag T is considered as remote tag by process P
+second time. (This only matters if Process P differentiates between remote and local tags. For e.g
+it ignores all remote tags and uses only local tags for aggregation. Note )
+
+
+A tag propagator should provide an option to declare that it is running on the edge of an entity.
```

Comment:
what is the use case that cares about "source"?

Reply:
Some process or micro-service could ignore remote tag for their metrics or logging or tracing but it would still forward the tags.

Reply:
That use case does not require pulling the "source" as a top-level concept, nor does the "source" address that use case fully. For example, I may want to set some tags in the application and propagate them down, yet I don't want them to be used as dimensions in the metrics. For example, I may want to propagate the UUID of the top-level user through the whole call graph, yet I'd never want to partition metrics by that value.

Reply:
I had a chat with @bogdandrutu and he had suggested view definition already takes care of whether to use the tag for metric partition or not. So it seems that all we need is
1. A scope to indicate that the tag is of local significance (process level) and should not be propagated.
2. Exclude/Include criteria for incoming and outgoing tags.